src/components/EditableField/EditableField.utils.ts

```diff
@@ -0,0 +1,78 @@
+import { FieldAction, FieldValue } from './EditableField.types'
+import { isArray, isObject } from '../../utilities/is'
+
+export const COMPONENT_KEY = 'EditableField'
+
+export const ACTION_ICONS = {
+  delete: 'cross-medium',
+  link: 'link',
+  plus: 'plus-medium',
+  valueOption: 'chevron-down',
+}
+
+export const deleteAction: FieldAction = {
+  name: 'delete',
+}
+
+export function normalizeFieldValue({
+  value,
+  name,
+  createNewFieldValue,
+  defaultOption,
+}): FieldValue[] {
+  let fieldValue: FieldValue[] = []
+
+  if (isArray(value)) {
+    fieldValue = value.map(val =>
+      createNewFieldValue({ value: val, name }, defaultOption)
+    )
+  } else {
+    fieldValue.push(createNewFieldValue({ value, name }, defaultOption))
+  }
+
+  return fieldValue
+}
+
+export function createNewValueFieldFactory(uuidFn) {
+  return function createNewValueFieldObject(
+    { value, name },
+    defaultOption?
+  ): FieldValue {
+    if (isObject(value)) {
+      const fieldObj = { ...value, id: uuidFn(`${name}_`) }
+
+      if (defaultOption != null && !Boolean(value.option)) {
+        fieldObj.option = defaultOption
+      }
+
+      return fieldObj
+    }
+    const fieldObj: any = { value, id: uuidFn(`${name}_`) }
+    if (defaultOption != null) {
+      fieldObj.option = defaultOption
+    }
+
+    return fieldObj
+  }
+}
+
+export function generateFieldActions(actions): FieldAction[] | [] {
+  let newActionsArray: FieldAction[] = []
+
+  if (actions !== null) {
+    if (actions === undefined) {
+      newActionsArray = [deleteAction]
+    } else {
+      newActionsArray = Array.isArray(actions) ? actions : [actions]
+
+      let isDeleteActionPresent =
+        newActionsArray.filter(action => action.name === 'delete').length > 0
+
+      newActionsArray = isDeleteActionPresent
+        ? newActionsArray
+        : newActionsArray.concat(deleteAction)
+    }
+  }
+
+  return newActionsArray
+}
```

Comment:
I'm finding this method a bit difficult to follow because of the nested `if` statements. 

I think we could employ early returns to make it all easier to follow what the final value being returned is.

```
if (actions === null) {
  return []
}

If (actions === undefined) {
  return [deletedAction]
}

if (!Array.isArray(actions)) {
  return [actions]
}

const isDeleteActionPresent = actions.filter(({ name }) => name === 'delete').length > 0
return isDeleteActionPresent
  ? actions
  : actions.concat(deleteAction)
```

Thoughts?

Reply:
As a follow-up question, is there a good use case for returning an empty set? I find it a bit unclear having different paths through code based on the difference between `null` and `undefined`. If there is a good case, could we add a comment?

Reply:
In this case I do need different handling of `null` and `undefined`, because by default the "delete" action gets added (so if actions is `undefined`, you get "delete") and if you explicitly want to remove this behaviour, you can pass `null`. I'll put a comment there.

I agree that it can be structured in a better manner though. Thanks!

Reply:
One more thing, you can also override the delete action, so I need to check for that too